### ISSUE TITLE
co_annotations branch caused a crash in stackeffect() in compile.c

### ISSUE THREAD

**Issue by larryhastings** [python/cpython#88057]
BPO | [43891](https://bugs.python.org/issue43891)
--- | :---
Nosy | @larryhastings

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-04-19.23:00:10.044>
labels = ['interpreter-core', '3.10', 'type-crash']
title = 'co_annotations branch caused a crash in stackeffect() in compile.c'
updated_at = <Date 2021-04-19.23:00:38.255>
user = 'https://github.com/larryhastings'
```

bugs.python.org fields:
```python
activity = <Date 2021-04-19.23:00:38.255>
actor = 'larry'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2021-04-19.23:00:10.044>
creator = 'larry'
dependencies = []
files = []
hgrepos = []
issue_num = 43891
keywords = []
message_count = 2.0
messages = ['391413', '391414']
nosy_count = 1.0
nosy_names = ['larry']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'crash'
url = 'https://bugs.python.org/issue43891'
versions = ['Python 3.10']
```

</p></details>

**Comment by iritkatriel** [comment]
Larry, I checked out your branch to test my hypothesis but I was unable to build it. If you still can build it, then perhaps you can try adding the assertion I suggest below to see if it trips.

The size of the stack is the number of blocks, and the intention is that each block is pushed at most once. The only way I can see blocks being pushed more than once is if in ``stackdepth_push`` we have ``b->b_startdepth < 0 && depth < 0 && b->b_startdepth < depth``.  By why would depth ever be <0?

It is possible that this happened if your changes caused the bytecode to be what we call "invalid", such that ``new_depth`` is negative. Unlike the other places where we calculate a depth, there is no assertion that new_depth is >= 0.   Then we have ``depth = new_depth``, and if we loop again we reach the assertion on ``depth >= 0``. But if this is the last iteration, then we don't check depth, we just call ``stackdepth_push(&sp, next, depth);`` after the loop ends.

If ``next`` is unvisited, then ``next->b_startdepth`` is ``MIN_INT``. So it will be smaller than some small-negative ``depth``. 

In summary, I think it is a bug in your code, which would have been easier to find if we had
``assert(new_depth >= 0); /* invalid code or bug in stackdepth() */`` just after new_depth is calculated.

I will make a patch to add this assertion.

**Comment by sweeneyde** [comment]
Could it be related to the discussion [here](https://github.com/python/cpython/pull/92769#discussion_r884981394) about potentially switching "visited" to mean "pushed" instead of the current meaning, "popped"? This could cause overflow if multiple instructions have the same target, and so things get pushed twice before they're popped.

**Comment by sweeneyde** [comment]
Never mind, I don't think that's the issue -- `stackdepth` already uses "visited"="pushed".

**Comment by iritkatriel** [comment]
That was my first thought as well, but here depth >= 0 seems to be used as a proxy for ‘visited’. So I think the only way we could insert something more than once is if its  depth is negative, but not MIN_INT.
